2019-11-auth-server

peeech opened the pull request:
Implements a server that verifies signatures created by HP Admin UI

peeech committed:
set up environment

peeech committed:
implement hyper http server

peeech committed:
extract body from request future

peeech committed:
add response with verification status

peeech committed:
construct and serialize payload for signature verification

peeech committed:
read X-Hpos-Admin-Signature header

peeech committed:
verify signatures pass one

peeech committed:
make use of X-Forwarded-For header as uri

peeech committed:
add read_hp_pubkey() WIP

peeech committed:
parse hp admin pubkey

peeech committed:
decode signature header from base64

peeech committed:
update headers to comply with documentation

peeech committed:
Update README.md

mjbrisebois commented:
Are there unit tests, or e2e tests?

peeech commented:
@mjbrisebois Entire suite of tests is intended to be developed together with `client` part. e2e in NixOS, then also unit tests can be written once we know the signature-payload combo that would produce `200 OK` response. For now your chances of getting `200 OK` are `1/2^256`. Check out the client story on PiT.

pjkundert commented on server/Cargo.toml:
```diff
@@ -0,0 +1,14 @@
+futures = "0.1.25"
+serde_json = "1.0.39"
+lazy_static = "1.2"
+ed25519-dalek = { version = "1.0.0-pre.1", features = ["nightly", "serde"] }
```
```suggestion
ed25519-dalek = { version = "1.0.0-pre.2", features = ["nightly", "serde"] }
```
We've moved on to pre.2 in other repos...

peeech replied:
fixed

pjkundert commented on server/src/main.rs:
```diff
@@ -0,0 +1,139 @@
+}
+
+fn verify_request(payload: String, headers: HeaderMap<HeaderValue>) -> bool {
+    // Retrieve X-Hpos-Admin-Signature, 401 on error
```
```suggestion
    // Retrieve X-Hpos-Admin-Signature, false on error
```

We're just returning the false here; it looks like `create_response` actually returns an  `200: OK` with a `false` body payload...

peeech replied:
clarified reasoning

pjkundert commented on server/src/main.rs:
```diff
@@ -0,0 +1,139 @@
+        None => return false,
+    };
+
+    // Base64 decode signature, 401 on error
```
```suggestion
    // Base64 decode signature
```

peeech replied:
fixed

pjkundert commented on server/Cargo.toml:
```diff
@@ -0,0 +1,14 @@
+edition = "2018"
+
+[dependencies]
+hpos-state-core = { git = "https://github.com/Holo-Host/hpos-state", branch = "develop" }
```
We won't be able to build this in Nix using this kind of incantation, right, @yegortimoshenko?

samrose replied:
I as able to make this work past the rust build part, but still other build issues with this PR


https://github.com/Holo-Host/hp-admin-crypto/pull/3

I will work on it more today and keep my work in the https://github.com/Holo-Host/hp-admin-crypto/pull/3

@peeech @pjkundert @yegortimoshenko

peeech replied:
Great, thanks!

pjkundert reviewed (dismissed):
Mostly great, I think!  @yegortimoshenko -- how do we build this in `holo-nixpkgs`, with the `Cargo.toml` https:// reference to `hpos-state`?

pjkundert commented on server/src/main.rs:
```diff
@@ -0,0 +1,139 @@
+}
+
+// Create response based on the request parameters
+fn crete_response(req: Request<Body>) -> impl Future<Item = Response<Body>, Error = hyper::Error> {
```
```suggestion
fn create_response(req: Request<Body>) -> impl Future<Item = Response<Body>, Error = hyper::Error> {
```

peeech replied:
fixed

pjkundert commented on server/src/main.rs:
```diff
@@ -0,0 +1,139 @@
+
+    // Create a `Service` from servicing function
+    let new_svc = || {
+        service::service_fn(crete_response)
```
```suggestion
        service::service_fn(create_response)
```

peeech replied:
fixed

pjkundert committed:
Add some Rust based unit tests
o Reorganize a bit to support Rust test of very_request
o Simplify logic of verify_request a bit to make return flow clearer
o Rename crete_response to create_response

peeech committed:
update ed25519-dalek to 1.0.0-pre.2

peeech committed:
fix typos and comments

peeech committed:
Merge branch '2019-11-auth-server' into 2019-11-auth-server-rust-tests

peeech committed:
Merge pull request #2 from pjkundert/2019-11-auth-server-rust-tests

Add Rust based unit tests for verify_request()

samrose committed:
server: fix Nix build

lukateras commented on .gitignore:
```diff
@@ -0,0 +1,5 @@
+pkg/
+result
+target/
+*/Cargo.lock
```
https://doc.rust-lang.org/cargo/faq.html#why-do-binaries-have-cargolock-in-version-control-but-not-libraries

```suggestion
```

peeech replied:
:+1:

lukateras commented on .gitignore:
```diff
@@ -0,0 +1,5 @@
+node_modules/
```
```suggestion
```

lukateras commented on .gitignore:
```diff
@@ -0,0 +1,5 @@
+node_modules/
+pkg/
```
```suggestion
```

lukateras commented on default.nix:
```diff
@@ -0,0 +1,21 @@
+
+{
+  hp-admin-authorize = buildRustPackage rustPlatform {
+    name = "hp-admin-authorize";
```
Dervation name doesn't match repo name or subdir name.

peeech replied:
resolved in #3

lukateras commented on default.nix:
```diff
@@ -0,0 +1,21 @@
+
+let
+  inherit (rust.packages.nightly) rustPlatform;
+  inherit (darwin.apple_sdk.frameworks) CoreServices Security;
```
`CoreServices` is imported, but never used.

peeech replied:
resolved in #3

lukateras commented on server/Cargo.toml:
```diff
@@ -0,0 +1,14 @@
+serde_json = "1.0.39"
+lazy_static = "1.2"
+ed25519-dalek = { version = "1.0.0-pre.2", features = ["nightly", "serde"] }
+base64 = "0.10.1"
```
Missing newline at EOL.

lukateras commented on server/src/main.rs:
```diff
@@ -0,0 +1,156 @@
+
+lazy_static! {
+    static ref X_HPOS_ADMIN_SIGNATURE: HeaderName = HeaderName::from_lowercase(b"x-hpos-admin-signature").unwrap();
+    static ref X_ORIGINAL_URI: HeaderName = HeaderName::from_lowercase(b"x-original-uri").unwrap();
```
This can only be URL rather than more generic URI, so it would be better to be more specific here:

```suggestion
    static ref X_ORIGINAL_URL: HeaderName = HeaderName::from_lowercase(b"x-original-url").unwrap();
```

peeech replied:
Well, nginx is capable of operating on more general notion of URIs and `X_ORIGINAL_URI` is a standard header in traffic proxying. I do not see the reason for semantically limiting our functionality. That said I'd much rather stay with `X_ORIGINAL_URI` header here.

lukateras replied:
Well, I've seen `x-original-url` being used more frequently than `x-original-uri`, and Google results seem to back it up: https://www.google.com/search?q=%22x-original-url%22 (92k results) vs https://www.google.com/search?q=%22x-original-uri%22 (5.4k results). It doesn't matter much either way, so nevermind.

lukateras commented on server/src/main.rs:
```diff
@@ -0,0 +1,156 @@
+        base64::encode_config_buf(signature.to_bytes().as_ref(), base64::STANDARD, &mut signature_base64);
+
+        let mut headers = HeaderMap::new();
+        headers.insert( "x-hpos-admin-signature", signature_base64.parse().unwrap() );
```
```suggestion
        headers.insert(X_HPOS_ADMIN_SIGNATURE, signature_base64.parse().unwrap() );
```

pjkundert replied:
Tried that, couldn't figure out how to make it work...

samrose replied:
I was able to get this project to build in nix, when I run it, I get the error:

```
result/bin/hp-admin-crypto-server 
thread 'main' panicked at 'HPOS_STATE_PATH environmental variable is not present: NotPresent', src/libcore/result.rs:1051:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

I was able to add this env var in the development context (for local machine development) to the shell.nix file like this

```
{ pkgs ? import ./pkgs.nix {} }:

with pkgs;

mkShell {
  inputsFrom = lib.attrValues (import ./. { inherit pkgs; });
  HPOS_STATE_PATH = "blah";
}
```

then running `nix-shell`

the variable is found 

```
 /nix/store/07cqdikifgsn0krv30ds3x2dc1rhvg9v-hp-admin-authorize/bin/hp-admin-crypto-server
Reading HP Admin Public Key from blah.
thread 'main' panicked at 'Something went wrong reading HP Admin Public Key from file: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/libcore/result.rs:1051:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

```

samrose replied:
I pushed this to PR #3

lukateras commented on server/src/main.rs:
```diff
@@ -0,0 +1,156 @@
+
+        let signature = secret_key_exp.sign( body_json.as_bytes(), &public_key );
+        let mut signature_base64 = String::new();
+        base64::encode_config_buf(signature.to_bytes().as_ref(), base64::STANDARD, &mut signature_base64);
```
```suggestion
        base64::encode_buf(signature.to_bytes().as_ref(), &mut signature_base64);
```

peeech replied:
this change === decision that we are using base64::STANDARD with padding

peeech replied:
Actually we are using base64::STANDARD_NO_PAD in verify_request(), so it should be `base64::encode_config_buf( ... , base64::STANDARD_NO_PAD, ... )`

yegortimoshenko commented on server/src/main.rs:
```diff
@@ -0,0 +1,156 @@
+        let public_key = ed25519_dalek::PublicKey::from( &secret_key );
+        let secret_key_exp = ed25519_dalek::ExpandedSecretKey::from( &secret_key );
+
+        // Now lets sign some body payload
```
```suggestion
```

lukateras commented on server/src/main.rs:
```diff
@@ -0,0 +1,156 @@
+
+        // Now lets sign some body payload
+        let body = json!({
+            "something": "interesting"
```
This seems to be not what it's supposed to be?

pjkundert replied:
We're just checking signing, not the body payload...

peeech replied:
Removed confusion here by renaming `body -> payload`

lukateras commented on server/src/main.rs:
```diff
@@ -0,0 +1,156 @@
+    #[test]
+    fn verify_request_smoke() {
+        // Get a legit request_hash signature, agent_id
+        let secret: [u8; 32] = [0_u8; 32];
```
Type signature is not needed here, also this should use [`SECRET_KEY_LENGTH`](https://docs.rs/ed25519-dalek/1.0.0-pre.1/ed25519_dalek/constant.SECRET_KEY_LENGTH.html).

```suggestion
        let secret = [0_u8; SECRET_KEY_LENGTH];
```

yegortimoshenko commented on server/src/main.rs:
```diff
@@ -0,0 +1,156 @@
+    fn verify_request_smoke() {
+        // Get a legit request_hash signature, agent_id
+        let secret: [u8; 32] = [0_u8; 32];
+        let secret_key = ed25519_dalek::SecretKey::from_bytes( &secret ).unwrap();
```
This should be properly error handled.

lukateras commented on server/src/main.rs:
```diff
@@ -0,0 +1,156 @@
+        // Get a legit request_hash signature, agent_id
+        let secret: [u8; 32] = [0_u8; 32];
+        let secret_key = ed25519_dalek::SecretKey::from_bytes( &secret ).unwrap();
+        let public_key = ed25519_dalek::PublicKey::from( &secret_key );
```
Run `cargo fmt`.

peeech replied:
Fixing this per [Rust style guide](https://github.com/rust-dev-tools/fmt-rfcs/blob/master/guide/expressions.md#function-calls)

lukateras commented on server/src/main.rs:
```diff
@@ -0,0 +1,156 @@
+        let secret: [u8; 32] = [0_u8; 32];
+        let secret_key = ed25519_dalek::SecretKey::from_bytes( &secret ).unwrap();
+        let public_key = ed25519_dalek::PublicKey::from( &secret_key );
+        let secret_key_exp = ed25519_dalek::ExpandedSecretKey::from( &secret_key );
```
If you want to sign, you should create `Keypair` rather than `ExpandedSecretKey`.

peeech replied:
From what I can see this is the easiest way to create signature

peeech replied:
In [this code](https://github.com/Holo-Host/hp-admin-crypto/blob/2019-12-05-keypair-test/server/src/main.rs#L149) it is apparent that the method using `ExpandedSecretKey` both works and is less complex.

yegortimoshenko commented on server/src/main.rs:
```diff
@@ -0,0 +1,156 @@
+        });
+        let body_json = serde_json::to_string( &body ).unwrap();
+
+        let signature = secret_key_exp.sign( body_json.as_bytes(), &public_key );
```
Server should only verify, why does it sign?

lukateras commented on server/src/main.rs:
```diff
@@ -0,0 +1,156 @@
+        let body = json!({
+            "something": "interesting"
+        });
+        let body_json = serde_json::to_string( &body ).unwrap();
```
I would much prefer if we handled errors in tests rather than unwrapping.

```suggestion
        let body_json = serde_json::to_string( &body )?;
```

peeech replied:
Why? This can't fail. If it does then this test by all means SHOULD panic!. Plus ? would exit test function with the type `serde_json::Result<T, Error>` while the return value is ().

lukateras commented on server/src/main.rs:
```diff
@@ -0,0 +1,156 @@
+}
+
+fn read_hp_pubkey() -> PublicKey {
+    let hpos_state_path = env::var("HPOS_STATE_PATH").expect("HPOS_STATE_PATH environmental variable is not present");
```
Function should return `Result<PublicKey, Error>` and allow caller to handle this error. This applies to all unwraps/excepts in this function.

lukateras commented on server/src/main.rs:
```diff
@@ -0,0 +1,156 @@
+    }); 
+
+    // Serialize it to a JSON string.
+    serde_json::to_string(&d).unwrap()
```
Same here: this can fail, return `Result`, unwrapping this inline is unsafe.

lukateras commented on server/src/main.rs:
```diff
@@ -0,0 +1,156 @@
+
+fn verify_request(payload: String, headers: HeaderMap<HeaderValue>, public_key: &PublicKey) -> bool {
+    if let Some(signature_base64) = headers.get(&*X_HPOS_ADMIN_SIGNATURE) {
+        if let Ok(signature_vec) = decode_config(&signature_base64, base64::STANDARD_NO_PAD) {
```
It's not clear what `decode_config` means here. 
```suggestion
        if let Ok(signature_vec) = base64::decode_config(&signature_base64, base64::STANDARD_NO_PAD) {
```

pjkundert replied:
I agree; was just trying to follow established convention...

lukateras commented on server/src/main.rs:
```diff
@@ -0,0 +1,156 @@
+
+fn main() {
+    // Listen on http socket port 2884 - "auth" in phonespell
+    let listen_address = ([127,0,0,1], 2884).into();
```
Let's use Unix sockets for new services.

peeech replied:
I am implementing http according to our conversation. As a reminder- accessing this service is not any usable attack vector. Using http allows on contrary to finish this task in the reasonable timeframe.

lukateras commented on server/src/main.rs:
```diff
@@ -0,0 +1,156 @@
+    // Listen on http socket port 2884 - "auth" in phonespell
+    let listen_address = ([127,0,0,1], 2884).into();
+
+    // Trigger lazy static to see if HP_PUBLIC_KEY assignment creates panic
```
It shouldn't be a constant if it can panic.

peeech replied:
Not really. The whole purpose of `lazy static` is that it allocates static reference to the result of function execution in the run-time. By default ANY run-time code execution can panic.

lukateras commented on server/src/main.rs:
```diff
@@ -0,0 +1,156 @@
+fn read_hp_pubkey() -> PublicKey {
+    let hpos_state_path = env::var("HPOS_STATE_PATH").expect("HPOS_STATE_PATH environmental variable is not present");
+
+    println!("Reading HP Admin Public Key from {}.", hpos_state_path);
```
At the very least:

```suggestion
    eprintln!("Reading HP Admin Public Key from {}.", hpos_state_path);
```

Better yet, this should use `log` crate.

peeech replied:
True, we want this to be an error message and go to `stderr`, not `stdout`

peeech replied:
Actually no - this conversation only makes sense in the context: what do we log, where do we log and what for do we log.

lukateras commented on server/src/main.rs:
```diff
@@ -0,0 +1,156 @@
+}
+
+fn create_payload (method: String, uri: String, body_string: String) -> String {
+    let d = json!({
```
This should be a struct of its own rather than set inline.

peeech replied:
done

peeech committed:
add verify_request_fail test

peeech committed:
fix formatting

samrose committed:
shell: set HPOS_STATE_PATH to vendored hpos-state.json

Co-authored-by: Yegor Timoshenko <yegortimoshenko@transumption.com>

peeech committed:
Move to struct Payload

peeech committed:
add logging

peeech committed:
move error handling to log and Result

peeech committed:
rebuild read_hp_pubkey to use refCell WIP

peeech committed:
remove all sources of panic

peeech committed:
global HP_PUBLIC_KEY as lazy_static + Mutex

peeech committed:
remove unnecessary error handling

peeech committed:
fix tests

lukateras committed:
hp-admin-authorize -> hp-admin-crypto-server

lukateras committed:
server: re-enable Nix build tests

lukateras committed:
server: alphabetically sort Cargo.toml

lukateras committed:
Remove unused CoreServices in root default.nix

lukateras committed:
Merge pull request #3 from Holo-Host/2019-11-auth-server-buildfixes

server: adding lockfile etc to get build working

lukateras committed:
server: run cargo fmt

peeech committed:
update README.md

peeech committed:
update test identity to known test credentials

lukateras reviewed (approved):
There are quite a few problems remaining but I'm going to merge as-is for the demo and fix this up later.
